### PR TITLE
[FLINK-12199][network] Refactor IOMetrics to not distinguish between local/remote in/out bytes

### PR DESCRIPTION
**Pull request by zhijiangW**
## What is the purpose of the change

*At the moment we have `numBytesInLocal/numBytesInRemote` in `IOMetrics` but in the actual usage we just always sum them up. As local/remote is the detail of shuffle service implementation, we can merge the local/remote metrics into one and increment it later in `Task` instead of remote/local channels in network internals.*

## Brief change log

  - *Remove legacy metric related codes*
  - *Merge local/remote related metrics into one*
  - *Modify `LocalInputChannel` and `RemoteInputChannel` to reference with the same metric*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zentol** [comment]
Even if _we_ combine them for the purpose of the UI, this may not be the case for users.

As it stands I'm against this change.

**Comment by zentol** [comment]
Basically, this change should be discussed on the ML first to gauge whether any user relies on these metrics to exist.

**Comment by zhijiangW** [comment]
Thanks for reviews and concerns @zentol .
I checked the previous usages of `IOMetrics#getNumBytesInRemote` and `IOMetrics#getNumBytesInLocal`, it seems that all the usages are merging these metrics together in handlers, so the framework does not provide the way for showing them separately ATM. Otherwise the users extend the handler to fetch separate metrics, but if in this case should we confirm the compatibility as before?

In flink framework we have not changed any behaviors as before, not functional loss. But we should not be responsible for the private version which is maintained by outside users. Or I missed anything usages atm?

**Comment by zhijiangW** [comment]
After further thinking through this issue, the current created local/remote counter would be registered into `MetricRegistry` and then notified for `MetricReporter`. If users have other `MetricReporter` implementations which might rely on these reported local/remote metrics, it seems better to keep the previous behavior although there are no separate usages of these metrics in flink framework handlers.

In order not to block the following work for decoupling the `TaskIOMetric` with `NetworkEnvironment`, I think of another way to work around. We could create the `numBytesInTotal` counter which seems more general for all the implementations in `TaskIOMetricGroup` instead of separate `numBytesInLocal` and `numBytesInRemote`. And during creating `SingleInputGate` we could pass `ProxyMetricGroup` for the input channel to create `numBytesInLocal` and `numBytesInRemote` separately, because these two metrics are specific with implementations of `LocalInputChannel` and `RemoteInputChannel`.

During `InputChannel#getNextBuffer`, the `numBytesInTotal` and `numBytesInLocal/Remote` are both updated internally. The current flink handlers could fetch `numBytesInTotal` directly via `IOMetrics`, and outside users could still see the meaningful `numBytesInLocal/Remote` metrics in specific reporter as before. What do you think ? @zentol  @azagrebin

**Comment by azagrebin** [comment]
@zhijiangW
The idea looks good to me. Only one thing to clarify: `during creating SingleInputGate we could pass ProxyMetricGroup for the input channel to create numBytesInLocal and numBytesInRemote`. At the moment, we have only one `numBytesInLocal/Remote` per task, so they have to be created only once per task. But I think we could already introduce the totals here for `IOMetrics` (also one per task) as you wrote and increment them in `SingleInputGate` before we introduce `InputGateWithMetrics` and think about `NetworkEnvironment#registerMetrics`.

**Comment by zhijiangW** [comment]
@zentol @azagrebin The codes are already updated for addressing above comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 1]
```diff
@@ -108,7 +110,7 @@ public RemoteInputChannel(
 		ResultPartitionID partitionId,
 		ConnectionID connectionId,
 		ConnectionManager connectionManager,
-		TaskIOMetricGroup metrics) {
+		ProxyMetricGroup metrics) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: I would suggest creating `TaskNetworkMetrics` internally in for `NetworkEnviroment` (now in `registerTask`) and create/register all metrics there, similar to how it is done in `TaskIOMetricGroup`. Then pass only what is needed to gate and channels. This way we could keep metrics concern separately. All aggregating gauges from `TaskIOMetricGroup.initializeBufferMetrics` could be also moved from `TaskIOMetricGroup` into `TaskNetworkMetrics` (separate commit or PR for FLINK-12202).

**Comment by zhijiangW** [comment]
Thanks for the reviews @azagrebin !

If I understand correctly, the total IO related metrics could be divided into two groups. One group is general and suitable for all `ShuffleService` implementations such as current bytesIn/Out, recordsIn/Out which would be still created in current `TaskIOMetrics`.

Another group is special for current `NetworkEnvironment` implementation. We could create two private classes `NetworkInput/OutputMetrics` inside `NetworkEnvironment`, and the current metrics of `numBytesInLocal/Remote`, `numBuffersInLocal/Remote`, `input/outputQueueLength`, `in/outPoolUsage` could be migrated from `TaskIOMetricGroup` into `NetworkInput/OutputMetrics`, then we could pass these special metrics into local/remote channels.

Maybe we could not need the new introduced `InputGateWithMetrics` future. We could pass general `Counter`  (`numBytesIn`, `numBuffersIn`) from `TaskIOMetrics` in the constructor of `SingleInputGate` to update these values properly, not pass specific `TaskIOMetricGroup` in the constructor. What do you think?

**Comment by zhijiangW** [comment]
I have updated the codes for reviews. @zentol  @azagrebin

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java [hunk 2]
```diff
@@ -150,7 +154,7 @@ public static RemoteInputChannel createRemoteInputChannel(
 			connectionManager,
 			0,
 			0,
-			UnregisteredMetricGroups.createUnregisteredTaskMetricGroup().getIOMetricGroup());
+			new InputChannelMetrics(UnregisteredMetricGroups.createUnregisteredTaskMetricGroup().getIOMetricGroup()));
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java]: `InputChannelTestUtils.newUnregisteredInputChannelMetrics()`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java [hunk 3]
```diff
@@ -93,6 +98,62 @@ public static ConnectionManager createDummyConnectionManager() throws Exception
...
+			new TaskEventDispatcher(),
+			0,
+			0,
+			UnregisteredMetricGroups.createUnregisteredTaskMetricGroup().getIOMetricGroup());
+	}
+
+	public static LocalInputChannel createLocalInputChannel(
+		SingleInputGate inputGate,
+		ResultPartitionManager partitionManager,
+		Tuple2<Integer, Integer> initialAndMaxRequestBackoff) {
```
[the remark this answers is not in the document]
Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java]: I just migrated it from previous `LocalInputChannelTest#createLocalInputChannel` to keep the same form. I think the separate parameters seem better. I would adjust it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java [hunk 4]
```diff
@@ -93,6 +98,62 @@ public static ConnectionManager createDummyConnectionManager() throws Exception
...
+	public static RemoteInputChannel createRemoteInputChannel(
+		SingleInputGate inputGate,
+		int channelIndex,
+		ConnectionManager connectionManager) {
+
+		return new RemoteInputChannel(
+			inputGate,
+			channelIndex,
+			new ResultPartitionID(),
+			mock(ConnectionID.class),
```
[the remark this answers is not in the document]
Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java]: Yes, it is better to not use mock here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 5]
```diff
@@ -666,7 +674,7 @@ public static SingleInputGate create(
 		NetworkEnvironment networkEnvironment,
 		TaskEventPublisher taskEventPublisher,
 		TaskActions taskActions,
-		TaskIOMetricGroup metrics) {
+		TaskIOMetricGroup parent) {
```
[the remark this answers is not in the document]
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: Yes, you are right. I have not thought of this issue before.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java (hunk 2 above)
Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java]: I guess we define a new method in `InputChannelTestUtils` for creating this `InputChannelMetrics` in order to reuse in multiple places?

Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java]: I suggest so

**Comment by zhijiangW** [comment]
Thanks for further review @azagrebin .
I have pushed one fixup commit for addressing above comments.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java [hunk 6]
```diff
@@ -179,15 +181,17 @@ public void testBackwardsEventWithUninitializedChannel() throws Exception {
 
 		inputGate.setBufferPool(bufferPool);
 
+		final InputChannelMetrics metrics = new InputChannelMetrics(UnregisteredMetricGroups.createUnregisteredTaskMetricGroup().getIOMetricGroup());
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java]: `InputChannelTestUtils.newUnregisteredInputChannelMetrics()`?

Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java]: fixed and squashed with last commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 7]
```diff
@@ -561,6 +567,7 @@ public void requestPartitions() throws IOException, InterruptedException {
 		}
 
 		final Buffer buffer = result.get().buffer();
+		numBytesIn.inc(buffer.getSizeUnsafe());
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: here we now count more stuff, but aren't counting less anywhere else. Why is that?

Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: This should be just yet another new total counter for all tasks (local + remote channels). Old counters with local and remote separation should stay as internal specific metrics of network environment. At the end of the day, total = local + remote and used directly without summing of specific counters outside of network environment in IOMetrics.

Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: Later we plan to move the total counter completely out of network environment to count it always in gate wrapper to make it non-shuffle specific: https://issues.apache.org/jira/browse/FLINK-12201

Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: exactly, thanks for replies. @azagrebin

Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: We add a new total metric used for all the existing handlers which keep the same behavior as before, then the existing local/remote metrics inside `NetworkEnvironment` could also be visible for outside users.

**Pull request merged**